File: lib/face_native/android/src/main/kotlin/com/example/face_native/FaceNativePlugin.kt
package com.example.face_native


import android.content.Context
import android.graphics.BitmapFactory
import androidx.annotation.NonNull
import com.example.face_native.data.ObjectBoxStore
import com.example.face_native.domain.ImageVectorUseCase

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import androidx.core.net.toUri
import android.util.Log
import androidx.exifinterface.media.ExifInterface
import com.example.face_native.data.FaceImageRecord
import com.example.face_native.di.AppModule

import org.koin.android.ext.koin.androidContext
import org.koin.core.context.startKoin
import org.koin.java.KoinJavaComponent.inject
import java.io.ByteArrayInputStream


class FaceNativePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
    private val imageVectorUseCase: ImageVectorUseCase by inject(clazz = ImageVectorUseCase::class.java)
    private lateinit var context: Context

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
      context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "face_native")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
   
    when (call.method) {
      "initObjectBox" ->{
          val tenantKey = call.argument<String>("tenantKey")
          if (tenantKey != null) {
              val initSuccess = ObjectBoxStore.init(context, tenantKey)
              if (initSuccess) {
                  // Only initialize Koin if ObjectBox init was successful and Koin is not already started
                  try {
                      if (org.koin.core.context.GlobalContext.getOrNull() == null) {
                          startKoin {
                              androidContext(context)
                              modules(AppModule().module)
                          }
                      }
                      result.success(true)
                  } catch (e: Exception) {
                      Log.e("MainActivity", "Failed to initialize Koin", e)
                      result.error("INIT_KOIN_ERROR", "Failed to initialize dependency injection: ${e.message}", null)
                  }
              } else {
                  result.error("INIT_OBJECT_BOX_ERROR", "Failed to initialize ObjectBox", null)
              }
          } else {
              result.error("INIT_OBJECT_BOX_ERROR", "Tenant key is required", null)
          }
      }
      "getFaceImageRecordByListEmpId" -> {
          val empIdList = call.argument<List<Long>>("empIdList")
          if (empIdList != null) {
              val records = imageVectorUseCase.getFaceImageRecordByListEmpId(empIdList)
              val recordsMap = records.map { record ->
                  mapOf(
                      "empId" to record.empId,
                          "personName" to record.personName,
                          "faceEmbedding" to record.faceEmbedding?.toList()
                  )
              }
              result.success(recordsMap)
          } else {
              result.error("GET_FACE_IMAGE_RECORD_BY_LIST_EMP_ID_ERROR", "EmpId list is required", null)
          }
      }
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "getCount" -> {
          try {
              val count = imageVectorUseCase.getCount()
              result.success(count)
          } catch (e: Exception) {
              result.error("GET_COUNT_ERROR", e.message, null)
          }
      }
      "addAllRecords" -> {
          try {
              val imageMaps = call.argument<List<Map<String, Any?>>>("records")
              val records = imageMaps
                  ?.mapNotNull { map ->
                      val empId = (map["empId"] as? Number)?.toLong() ?: 0L
                      val personName = map["personName"] as? String
                      val faceEmbeddingList = map["faceEmbedding"] as? List<*>
                      val faceEmbedding = faceEmbeddingList
                          ?.mapNotNull { it as? Number }
                          ?.map { it.toFloat() }
                          ?.toFloatArray()
                      if (personName != null && faceEmbedding != null) {
                          FaceImageRecord(
                              empId = empId,
                              personName = personName,
                              faceEmbedding = faceEmbedding
                          )
                      } else {
                          null
                      }
                  }
              if (records != null) {
                  imageVectorUseCase.addAllRecords(records)
                  result.success(true)
              } else {
                  result.error("ADD_ALL_RECORDS_ERROR", "Invalid records data", null)
              }
          } catch (e: Exception) {
              result.error("ADD_ALL_RECORDS_ERROR", e.message, null)
          }
      }
      "getAllImages" -> {
          try {
              val images = imageVectorUseCase.getAllImages()
              val imageList = images
                  .filter {  it.personName != null && it.faceEmbedding != null }
                  .map { image ->
                      mapOf(
                          "empId" to image.empId,
                          "personName" to image.personName,
                          "faceEmbedding" to image.faceEmbedding
                      )
                  }
              result.success(imageList)
          } catch (e: Exception) {
              result.error("GET_ALL_IMAGES_ERROR", e.message, null)
          }
      }
      "getDatabaseSizeInBytes" -> {
          try {
              val sizeInBytes = imageVectorUseCase.getDatabaseSizeInBytes()
              result.success(sizeInBytes)
          } catch (e: Exception) {
              result.error("GET_DATABASE_SIZE_ERROR", e.message, null)
          }
      }
      "clearAllImages" -> {
          try {
              imageVectorUseCase.clearAllImages()
              result.success(true)
          } catch (e: Exception) {
              result.error("CLEAR_IMAGES_ERROR", e.message, null)
          }
      }
      "addImage" -> {
          val empId = call.argument<Int>("empId")?.toLong() // Using empId but called personId for compatibility
          val personName = call.argument<String>("personName")
          val imageUri = call.argument<String>("imageUri")
Log.d("NativeDemo", "addImage called with empId=$empId, personName=$personName, imageUri=$imageUri")
          if (empId != null && personName != null && imageUri != null) {
              CoroutineScope(Dispatchers.IO).launch {
                  try {
                      val uri = imageUri.toUri()
                          val addResult = imageVectorUseCase.addImage(empId, personName, uri)
                      withContext(Dispatchers.Main) {
                          if (addResult.isSuccess) {
                              result.success(addResult.getOrNull())
                          } else {
                              result.error("ADD_IMAGE_ERROR", addResult.exceptionOrNull()?.message ?: "Unknown error in addImage", null)
                          }
                      }
                  } catch (e: Exception) {
                      withContext(Dispatchers.Main) {
                          result.error("ADD_IMAGE_ERROR", e.message ?: "Unknown error in addImage", null)
                      }
                  }
              }
          } else {
              result.error("INVALID_ARGUMENTS", "PersonId, personName, and imageUri are required", null)
          }
      }
      "recognizeFace" -> {
          val imageBytes = call.argument<ByteArray>("imageBytes")

          if (imageBytes != null) {
              CoroutineScope(Dispatchers.IO).launch {
                  try {
                      val inputStream = ByteArrayInputStream(imageBytes)
                      val exif = ExifInterface(inputStream)
                      inputStream.close()
                      val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)

                      val (metrics, recognitionResults) = imageVectorUseCase.getNearestPersonName(bitmap, exif)

                      // Convert results to map for Flutter
                      val resultsMap = recognitionResults.map { result ->
                          mapOf(
                              "employeeId" to result.empId,
                              "personName" to result.personName,
                              "boundingBox" to mapOf(
                                  "left" to result.boundingBox.left,
                                  "top" to result.boundingBox.top,
                                  "right" to result.boundingBox.right,
                                  "bottom" to result.boundingBox.bottom
                              ),
                              "spoofResult" to result.spoofResult?.let { spoof ->
                                  mapOf(
                                      "isSpoof" to spoof.isSpoof,
                                      "score" to spoof.score,
                                      "timeMillis" to spoof.timeMillis
                                  )
                              }
                          )
                      }

                      val responseMap = mapOf(
                          "results" to resultsMap,
                          "metrics" to metrics?.let { m ->
                              mapOf(
                                  "timeFaceDetection" to m.timeFaceDetection,
                                  "timeFaceEmbedding" to m.timeFaceEmbedding,
                                  "timeVectorSearch" to m.timeVectorSearch,
                                  "timeFaceSpoofDetection" to m.timeFaceSpoofDetection
                              )
                          }
                      )

                      withContext(Dispatchers.Main) {
                          result.success(responseMap)
                      }
                  } catch (e: Exception) {
                      withContext(Dispatchers.Main) {
                          result.error("RECOGNIZE_FACE_ERROR", e.message, null)
                      }
                  }
              }
          } else {
              result.error("INVALID_ARGUMENTS", "base64Image is required", null)
          }
      }
      "removeImages" -> {
          val empId = call.argument<Int>("empId")?.toLong() // Using empId but called personId for compatibility

          if (empId != null) {
              try {
                  imageVectorUseCase.removeImages(empId)
                  result.success(true)
              } catch (e: Exception) {
                  result.error("REMOVE_IMAGES_ERROR", e.message, null)
              }
          } else {
              result.error("INVALID_ARGUMENTS", "PersonId is required", null)
          }
      }
      "getImageIdsByEmpId" -> {
          val empId = call.argument<Int>("empId")?.toLong()
          if (empId != null) {
              val imageIds = imageVectorUseCase.getImageIdsByEmpId(empId)
              result.success(imageIds)
          } else {
              result.error("INVALID_ARGUMENTS", "PersonId is required", null)
          }
        }
        
        "removeImagesByIds" -> {
          val imageIds = call.argument<List<Int>>("imageIds")
          val longImageIds = imageIds?.map { it.toLong() }
          if (longImageIds != null) {
              imageVectorUseCase.removeImagesByIds(longImageIds)
              result.success(true)
          } else {
              result.error("INVALID_ARGUMENTS", "ImageIds is required", null)
          }
        }
      else -> result.notImplemented()
  }



  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: lib/face_native/ios/Classes/FaceNativePlugin.swift
import Flutter
import UIKit

public class FaceNativePlugin: NSObject, FlutterPlugin {
  private let imageVectorUseCase: ImageVectorUseCase = ImageVectorUseCase.shared
    private let imagesVectorDB: ImagesVectorDB = ImagesVectorDB.shared
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "face_native", binaryMessenger: registrar.messenger())
    let instance = FaceNativePlugin()
    ObjectBoxStore.initialize(tenantKey: "test")
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

     func getUIImageFromPath(path: String) throws -> UIImage {
         guard let imageUrl = URL(string: path) else {
      print ("Error in getUIImageFromPath: Invalid image URL")
           throw FaceNetError.invalidImageUrl
         }
         if !FileManager.default.fileExists(atPath: imageUrl.path) {
             print("❌ File does not exist at path: \(imageUrl.path)")
          print("Error in getUIImageFromPath: File does not exist at path: \(imageUrl.path)")
             throw FaceNetError.invalidImageUrl
         }
         let image = UIImage(contentsOfFile: imageUrl.path)
         if image == nil {
          print("Error in getUIImageFromPath: Failed to create UIImage from data")
          throw FaceNetError.invalidImageUrl
         }
         return image!
        
    }
    
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
      
    case "initObjectBox":
      guard let args = call.arguments as? [String: Any],
            let tenantKey = args["tenantKey"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "TenantKey is required", details: nil))
        return
      }
      
     let isInitialized = ObjectBoxStore.initialize(tenantKey: tenantKey)
      if isInitialized {
        result(true)
      } else {
        result(FlutterError(code: "INIT_OBJECT_BOX_ERROR", message: "Failed to initialize ObjectBox with tenant key: \(tenantKey)", details: nil))
      }
      case "getFaceImageRecordByListEmpId" :
      guard let args = call.arguments as? [String: Any],
      let empIdList = args["empIdList"] as? [Int64] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "EmpId list is required", details: nil))
        return
      }
        do {
            let records = try imagesVectorDB.getFaceImageRecordByListEmpId(empIdList: empIdList)
            result(records.map { record in return
              [
                "empId": record.empId,
                "personName": record.personName,
                "faceEmbedding": record.faceEmbedding
              ]
            })
        }
        catch {
          result(FlutterError(code: "GET_FACE_IMAGE_RECORD_BY_LIST_EMP_ID_ERROR", message: error.localizedDescription, details: nil))
        }
    case "getCount" :
      do {
        let count = try imagesVectorDB.getCount()
        result(count)
      }
      catch {
        result(FlutterError(code: "GET_COUNT_ERROR", message: error.localizedDescription, details: nil))
      }
      case "addAllRecords" :
      guard let args = call.arguments as? [String: Any],
      let recordsMap = args["records"] as? [[String: Any]] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Records is required", details: nil))
        return
      }
      do {
          let filteredRecords = recordsMap.filter { record in
              (record["faceEmbedding"] as? [Float])?.isEmpty == false &&
              (record["empId"] as? Int64) != nil
          }

          let records = filteredRecords.map { record in
              FaceImageRecord(
                  empId: record["empId"] as! Int64,
                  personName: record["personName"] as? String ?? "",
                  faceEmbedding: record["faceEmbedding"] as! [Float]
              )
          }     
          try imagesVectorDB.addAllRecords(records: records)
        result(true)
      }
      catch {
          result(FlutterError(code: "ADD_ALL_RECORDS_ERROR", message: error.localizedDescription , details: nil))
      }
      case "getAllImages" :
      do {
     let records = try imagesVectorDB.getAllRecords()
    result(records.map { record in
      [
        "empId": record.empId,
        "personName": record.personName,
        "faceEmbedding": record.faceEmbedding
      ]
    })
      }
      catch {
        result(FlutterError(code: "GET_ALL_IMAGES_ERROR", message: error.localizedDescription, details: nil))
      }
      case "removeImages" :
      guard let args = call.arguments as? [String: Any],
      let empId = args["empId"] as? Int64 else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "EmpId is required", details: nil))
        return
      }
      do {
        try imagesVectorDB.removeFaceRecordsWithEmpId(empId: empId)
        result(true)
      }
      catch {
        result(FlutterError(code: "REMOVE_IMAGES_ERROR", message: error.localizedDescription, details: nil))
      }
      case "getImageIdsByEmpId" :
      guard let args = call.arguments as? [String: Any],
      let empId = args["empId"] as? Int64 else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "EmpId is required", details: nil))
        return
      }
      do {
        let ids = try imagesVectorDB.getImageIdsByEmpId(empId: empId)
        result(ids)
      }
      catch {
        result(FlutterError(code: "GET_IMAGE_IDS_BY_EMP_ID_ERROR", message: error.localizedDescription, details: nil))
      }
      case "removeImagesByIds" :
      guard let args = call.arguments as? [String: Any],
      let imageIds = args["imageIds"] as? [Int64] else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "ImageIds is required", details: nil))
        return
      }
      do {
        try imagesVectorDB.removeFaceRecordsByIds(imageIds: imageIds)
        result(true)
      }
      catch {       
        result(FlutterError(code: "REMOVE_IMAGES_BY_IDS_ERROR", message: error.localizedDescription, details: nil))
      }
    case "testGetCroppedFace":
        let testString="abc"
        print(testString)
      guard let args = call.arguments as? [String: Any],
            let imageUri = args["imageUri"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "ImageUri is required", details: nil))
        return
      }
        print(imageUri)
      Task {
        guard let imageUrl = URL(string: imageUri) else {
          result(FlutterError(code: "INVALID_URL", message: "Invalid image URL", details: nil))
          return
        }
        let faceDetector = MediapipeFaceDetector.shared
        let croppedFaceResult = await faceDetector.getCroppedFace(imageUrl: imageUrl)
        switch croppedFaceResult {
        case .success(let croppedImage):
          print("croppedImage success")
          result(true)
        case .failure(let error):
          result(FlutterError(code: error.code.rawValue, message: error.localizedDescription, details: nil))
        }
      }
    case "testGetFaceEmbedding":
      guard let args = call.arguments as? [String: Any],
            let imageUri = args["imageUri"] as? String else {
        result(FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "ImageUri is required",
          details: nil
        ))
        return
      }
        print("imageUri in testGetFaceEmbedding  \(imageUri)")
      Task {
        do {
          guard let imageUrl = URL(string: imageUri) else {
            result(FlutterError(
              code: "INVALID_URL",
              message: "Invalid image URL",
              details: nil
            ))
            return
          }
          let faceNet = FaceNet.shared
            print("🔍 Starting getCroppedFace with URL: \(imageUrl)")
            // Load image from URL
            if !FileManager.default.fileExists(atPath: imageUrl.path) {
                print("❌ File does not exist at path: \(imageUrl.path)")
                result(
                  FlutterError(
                    code: "FILE_NOT_FOUND",
                    message: "File does not exist at path: \(imageUrl.path)",
                    details: nil
                  ))
                return
            }
            
            guard let image = UIImage.init(contentsOfFile: imageUrl.path) else {
                print("❌ Failed to create UIImage from data")
                result(
                  FlutterError(
                    code: "FAILED_TO_CREATE_UIImage",
                    message: "Failed to create UIImage from data",
                    details: nil
                  ))
                return
            }
          let embedding = try await faceNet.getFaceEmbedding(image: image)
          result(embedding)
        } catch {
          result(FlutterError(
            code: "FACE_EMBEDDING_ERROR",
            message: "Failed to get face embedding: \(error.localizedDescription)",
            details: nil
          ))
        }
      }
    case "addImage" :
      guard let args = call.arguments as? [String: Any],
            let empId = args["empId"] as? Int64,
            let personName = args["personName"] as? String,
            let imageUri = args["imageUri"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "EmpId, personName, and imageUri are required", details: nil))
        return
      }
      Task {
        do {
//          let image = try getUIImageFromPath(path: imageUri)
          let recognitionResult =  await imageVectorUseCase.addImage(empId: empId, personName: personName, imageUri: imageUri)
          switch recognitionResult {
          case .success(let id):
            result(id)
          case .failure(let error):
            result(FlutterError(code: "ADD_IMAGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "recognizeFace" :
      guard let args = call.arguments as? [String: Any],
      let imageUrl = args["imageUrl"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "ImageUrl is required", details: nil))
        return
      }
      Task {
        do {
            let recognitionResult = await self.imageVectorUseCase.getNearestPersonName(imageUri: imageUrl)
          switch recognitionResult {
          case .success(let recognitionResult):
            print("recognitionResult in recognizeFace: \(recognitionResult.toMap())")
            result([
              "result": recognitionResult.toMap()
            ])
          case .failure(let error):
            result(FlutterError(code: "RECOGNIZE_FACE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
      case "getAllImages" :
      Task {
        do {
          let records = await self.imageVectorUseCase.getAllRecords()
          switch records {
          case .success(let records):
          print("records in getAllImages: \(records.map { $0.toMap() })")
            result(records.map { $0.toMap() })
          case .failure(let error):
          result(FlutterError(code: "GET_ALL_RECORDS_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
